document/encoding/msgpack/codec.go

```diff
@@ -0,0 +1,348 @@
+			return err
+		}
+
+		return e.EncodeValue(v)
+	})
+}
+
+// EncodeArray encodes a as a MessagePack array.
+func (e *Encoder) EncodeArray(a document.Array) error {
+	var alen int
+	var err error
+
+	vb, ok := a.(document.ValueBuffer)
+	if ok {
+		alen = len(vb)
+	} else {
+		alen, err = document.ArrayLength(a)
+		if err != nil {
+			return err
+		}
+	}
+
+	if err := e.enc.EncodeArrayLen(alen); err != nil {
+		return err
+	}
+
+	return a.Iterate(func(i int, v document.Value) error {
+		return e.EncodeValue(v)
+	})
+}
+
+// EncodeValue encodes v based on its type.
+// - document -> map
+// - array -> array
+// - NULL -> nil
+// - text -> string
+// - blob -> bytes
+// - bool -> bool
+// - int8 -> int8
+// - int16 -> int16
+// - int32 -> int32
+// - int64 -> int64
+// - float64 -> float64
+// - duration -> custom type with code 0x1 and size 8
+func (e *Encoder) EncodeValue(v document.Value) error {
+	switch v.Type {
+	case document.DocumentValue:
+		d, err := v.ConvertToDocument()
+		if err != nil {
+			return err
+		}
+		return e.EncodeDocument(d)
+	case document.ArrayValue:
+		a, err := v.ConvertToArray()
+		if err != nil {
+			return err
+		}
+		return e.EncodeArray(a)
+	case document.NullValue:
+		return e.enc.EncodeNil()
+	case document.TextValue:
+		s, err := v.ConvertToText()
+		if err != nil {
+			return err
+		}
+		return e.enc.EncodeString(s)
+	case document.BlobValue:
+		b, err := v.ConvertToBlob()
+		if err != nil {
+			return err
+		}
+		return e.enc.EncodeBytes(b)
+	case document.BoolValue:
+		return e.enc.EncodeBool(v.V.(bool))
+	case document.Int8Value:
+		return e.enc.EncodeInt8(v.V.(int8))
+	case document.Int16Value:
+		return e.enc.EncodeInt16(v.V.(int16))
+	case document.Int32Value:
+		return e.enc.EncodeInt32(v.V.(int32))
+	case document.Int64Value:
+		return e.enc.EncodeInt64(v.V.(int64))
+	case document.Float64Value:
+		return e.enc.EncodeFloat64(v.V.(float64))
+	case document.DurationValue:
+		err := e.enc.EncodeExtHeader(DurationType, 8)
+		if err != nil {
+			return err
+		}
+
+		d := uint64(v.V.(time.Duration))
+		var buf [8]byte
+		buf[0] = byte(d >> 56)
+		buf[1] = byte(d >> 48)
+		buf[2] = byte(d >> 40)
+		buf[3] = byte(d >> 32)
+		buf[4] = byte(d >> 24)
+		buf[5] = byte(d >> 16)
+		buf[6] = byte(d >> 8)
+		buf[7] = byte(d)
+
+		_, err = e.enc.Writer().Write(buf[:])
+		return err
+	}
+
+	return e.enc.Encode(v.V)
+}
+
+// Close puts the encoder into the pool for reuse.
+func (e *Encoder) Close() {
+	msgpack.PutEncoder(e.enc)
+}
+
+// Decoder decodes Genji documents and values
+// from MessagePack.
+type Decoder struct {
+	dec *msgpack.Decoder
+}
+
+// NewDecoder creates a Decoder that reads from the given reader.
+func NewDecoder(r io.Reader) *Decoder {
+	dec := msgpack.GetDecoder()
+	dec.Reset(r)
+
+	return &Decoder{
+		dec: dec,
+	}
+}
+
+// DecodeValue reads one value from the reader and decodes it.
+func (d *Decoder) DecodeValue() (v document.Value, err error) {
+	c, err := d.dec.PeekCode()
```

Comment:
We don't because almost every Decode* method of the decoder reads that code. Example: https://github.com/vmihailenco/msgpack/blob/e7759683b74a27e455669b525427cfd9aec0790e/decode_number.go#L83